firebase/rules/tests/invocations.spec.ts

```diff
@@ -0,0 +1,161 @@
+import 'jest';
+import * as matchers from '../matchers';
+import * as targaryen from 'targaryen';
+
+import { SYSTEM_USER, Invocation, InvocationType } from '@machinelabs/models';
+
+const anonymousUser = targaryen.util.users.anonymous;
+const rules = require('../database.rules.json');
+
+const dummyInvocation: Invocation = {
+  id: '1',
+  timestamp: Date.now(),
+  user_id: null,
+  type: Date.now() % 2 ? InvocationType.StartExecution : InvocationType.StopExecution,
+  data: null
+};
+
+const dummyInvocationLab = {
+  id: '1',
+  directory: 'dummy/directory'
+};
+
+const dummyInvocationExecution = {
+  execution_id: '1',
+};
+
+describe('/invocations', () => {
+  let currentUser;
+  let systemUser;
+
+  let dummyInvocationCurrentUser: Invocation;
+  let dummyInvocationSystemUser: Invocation;
+
+  beforeEach(() => {
+    jest.addMatchers(matchers);
+
+    targaryen.util.setFirebaseData({
+      users: {
+        '1': {
+          common: {
+            id: '1',
+            displayName: 'Foo Bar',
+            email: 'foo.bar@example.com',
+            bio: '',
+            isAnonymous: false,
+            photoUrl: 'avatars.com/1'
+          }
+        },
+        '2': {
+          common: {
+            id: '2',
+            displayName: 'Fizz Buzz',
+            email: 'fizz.buzz@example.com',
+            bio: '',
+            isAnonymous: false,
+            photoUrl: 'avatars.com/2'
+          }
+        }
+      },
+      invocations: {
+        '1': {
+          common: {
+            id: '1',
+            timestamp: Date.now(),
+            user_id: '1',
+            type: InvocationType.StartExecution,
+            data: dummyInvocationExecution
+          }
+        },
+        '3': {
+          common: {
+            id: '3',
+            timestamp: Date.now(),
+            user_id: '2',
+            type: InvocationType.StopExecution,
+            data: dummyInvocationLab
+          }
+        }
+      },
+      idx: {
+        invocation_rate_proof: {
+          '1': {
+            key: '1'
+          },
+          '2': {
+            key: '3'
+          }
+        }
+      }
+    });
+
+    currentUser = { uid: '1', id: '1', email: 'foo@bar.com', provider: 'github' };
+    systemUser = { uid: SYSTEM_USER };
+
+    dummyInvocationCurrentUser = { ...dummyInvocation, user_id: currentUser.uid };
+    dummyInvocationSystemUser = { ...dummyInvocation, user_id: SYSTEM_USER };
+
+    targaryen.util.setFirebaseRules(rules);
+  });
+
+  it('A user cannot read any root invocation', () => {
+    expect(currentUser).cannotRead('/invocations/1');
+    expect(currentUser).cannotRead('/invocations/3');
+  });
+
+  it('A user cannot write into a root invocation', () => {
```

Comment:
To elaborate on what @PascalPrecht said, the `system` being able to write into `/invocations/n/*` is an important part of the protocol because this is how we eventually assign a server to an invocation to take the code and run it.

After the cloud function picked up an incoming invocation and assigned a server it will look like this:

![image](https://user-images.githubusercontent.com/521109/38730537-2d2d49c6-3f17-11e8-8844-f304152c9d4e.png)